trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        >>> mf.fit_mf()
+
+    """
+
+    def __init__(  # pylint: disable-all
+        self,
+        num_users=None,
+        num_items=None,
+        num_latent_factors=None,
+        user_representation=None,
+        item_representation=None,
+        actual_user_representation=None,
+        actual_item_representation=None,
+        probabilistic_recommendations=False,
+        seed=None,
+        verbose=False,
+        num_items_per_iter=10,
+        model_params=None,
+        **kwargs
+    ):
+        # check inputs
+        num_users, num_items, num_attributes = validate_user_item_inputs(
+            num_users,
+            num_items,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            100,
+            1250,
+            10,
+            num_latent_factors,
+        )
+        if model_params is None:
+            model_params = {}
+        num_features_vals = non_none_values(
+            model_params.get("features"), num_latent_factors, num_attributes
+        )
+        if len(num_features_vals) > 1:
+            raise ValueError("Number of latent factors is not the same across inputs.")
+
+        self.num_latent_factors = num_attributes
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
+        if user_representation is None:
+            user_representation = np.zeros((num_users, num_attributes))
+        if item_representation is None:
+            item_representation = Generator(seed=seed).binomial(
+                n=1, p=0.5, size=(num_attributes, num_items)
+            )
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
```

Comment:
I thought we were trying to avoid dataframes and stick with numpy arrays? I'm totally fine with using dataframes, just want to make sure it doesn't cause any consistency problems...

Reply:
In this case, we can't avoid it because the `lenskit` function takes in a `DataFrame` as input, so we have to have a dataframe somewhere. There won't be any consistency problems because only the `ImplicitMF` class uses this internal dataframe to keep track of interactions. Good eye, though, it's important to make sure we don't introduce needless new data structures!

Reply:
Yes, but we could avoid having the pandas dataframe everywhere except for in the function that actually fits the LensKit model. That's why in my fit function I had these two lines

```
#create long dataframe with user id and item id of items users interacted with
nonzero_idx=np.nonzero(item_idxs)
idxs=tuple(zip(nonzero_idx[0], nonzero_idx[1]))
```
which would allow you to continue to store the user item interactions in the format I think they are stored everywhere else in t-recs, right? Not really a critical issue, but I wonder what @elucherini thinks!

Reply:
Yeah, I understand what you're saying, and I originally was using those two lines that you had before converting it to this other way. I just don't see a reason *not* to store interactions in a pandas dataframe all the time, because the MF model is the only model that we have that stores a history of interactions throughout multiple interactions. Furthermore, storing it in a numpy matrix that is U x I in dimension takes up more space (granted, we could use a sparse matrix, but that feels like it's part of a different PR) than just the pandas dataframe of (user,item) tuples. Since the interaction history is never used anywhere besides to pass into the `fit_mf` function, it seems unnecessary to store it as a numpy matrix first and then convert to a dataframe after.

Reply:
Do I understand correctly that the pandas dataframe is smaller than a dense numpy array because pandas represents dataframes internally as sparse matrices?
I do agree that a dense numpy array taking more memory than needed is not the best choice. I am also not a fan of using pandas dataframe internally, so I have a question: does `lenskit.algorithms.als` only work with pandas dataframes? Is it compatible with sparse matrices? If we can entirely skip the conversion to pandas dataframe, then I'd suggest using a sparse matrix. If that step is unavoidable, the current state is probably the best solution.

Reply:
I didn't try to test a numpy array because the LensKit documentation explicitly says that a pandas dataframe is required. I can file an issue with LensKit, but I don't expect that would get resolved in time for this merge :)